lang/java/compiler/src/main/javacc/org/apache/avro/compiler/idl/idl.jj

```diff
@@ -182,6 +183,32 @@ public class Idl implements Closeable {
     return result;
   }
 
+  /**
+   * For "optional schemas" (recognized by the marker property the NullableType
+   * production adds), ensure the null schema is in the right place.
+   *
+   * @param schema       a schema
+   * @param defaultValue the intended default value
+   * @return the schema, or an optional schema with null in the right place
+   */
+  private static Schema fixOptionalSchema(Schema schema, JsonNode defaultValue) {
```

Comment:
I like this so much!  This is a compelling feature... I'm tempted to say that nobody really **_wants_**
```
union { null, double } value = NaN
```
to fail either...  What would you think about auto-correcting all `union { null, notNull }` or `union { notNull, null } ` types that have a default?

I guess this might be too magical, and i suspect that the `?` syntax will be happily adopted in any case.

Reply:
Nothing too magical; I can check for schemas of null with non-null: `schema.isUnion() && schema.getTypes().size() == 2 && schema.getTypes().get(0).isNullable() != schema.getTypes().get(1).isNullable()`

And then swap the union if: `defaultValue != null && !defaultValue.isNull() && schema.getTypes().get(0).isNullable()`

However, I can also do this in the Schema.Field constructor.

Or even combine it with `org.apache.avro.Schema#isValidDefault(Schema, JsonNode)` to select the first schema for which the default is valid, and place it as the first schema in the union.

I've posted a poll to the dev mailinglist to choose between the options.

Reply:
The main problem is a) union instances must still have the default value match the first schema (I don't know enough to relax that constraint), and b) the union will not be what was written.

As a result, my preference is to restrict this feature to the `?` syntax.